### PR TITLE
Use a connection pool for gRPC client calls from jobservice to armada-server

### PR DESCRIPTION
**Pull request by ClifHouck**
Addresses [BATCH-225](https://gr-oss.atlassian.net/browse/BATCH-225) 

Uses my [custom forked version](https://github.com/ClifHouck/grpc-go-pool) of [grpc-go-pool](https://github.com/processout/grpc-go-pool) to add a pooled grpc armada-server client to the jobservice.

[BATCH-223]: https://gr-oss.atlassian.net/browse/BATCH-223?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

┆Issue is synchronized with this [Jira Task](https://gr-oss.atlassian.net/browse/BATCH-225) by [Unito](https://www.unito.io)

[BATCH-225]: https://gr-oss.atlassian.net/browse/BATCH-225?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: go.sum [hunk 1]
```diff
@@ -63,6 +63,8 @@ github.com/Azure/go-ntlmssp v0.0.0-20220621081337-cb9428e4ac1e h1:NeAW1fUYUEWhft
 github.com/Azure/go-ntlmssp v0.0.0-20220621081337-cb9428e4ac1e/go.mod h1:chxPXzSsl7ZWRAuOIE23GDNzjWuZquvFlgA8xmpunjU=
 github.com/BurntSushi/toml v0.3.1/go.mod h1:xHWCNGjB5oqiDr8zfno3MHue2Ht5sIBksp03qcyfWMU=
 github.com/BurntSushi/xgb v0.0.0-20160522181843-27f122750802/go.mod h1:IVnqGOEym/WlBOVXweHU+Q+/VP0lqqI8lqeDx9IjBqo=
+github.com/ClifHouck/grpc-go-pool v0.1.0 h1:YlG2qlfB/eEfM87Jv1FxfEeiuNiDFyABla0rRR4qFI4=
```
Comment (suprjinx) [inline go.sum:66]: should this be referencing your github?

**Review by suprjinx — APPROVED** [review]

File: go.sum (hunk 1 above)
Comment (dave-gantenbein) [reply go.sum:66]: +1

File: internal/jobservice/application.go [hunk 2]
```diff
@@ -30,12 +34,41 @@ func New() *App {
 	return &App{}
 }
 
+// Returns a non-nil error if mis-configuration is unrecoverable.
+func CheckConfig(config *configuration.JobServiceConfiguration) error {
```
Comment (dave-gantenbein) [inline internal/jobservice/application.go:38]: #testme

File: internal/jobservice/events/client.go
Comment (dave-gantenbein) [inline internal/jobservice/events/client.go:1]: Should we be testing the new methods here?

File: go.sum (hunk 1 above)
Comment (ClifHouck) [reply go.sum:66]: Depends. The root of my fork does not seem to be actively maintained. I've made modifications to the pool that make more sense for grpc connections. For instance, grpc connections are safe to use across many gorountines at once, so a client grabbing one shouldn't remove it from the pool. The pool just round-robins access to clients to load balance their use.

File: internal/jobservice/events/client.go
Comment (ClifHouck) [reply internal/jobservice/events/client.go:1]: Yes, will do.

File: go.sum (hunk 1 above)
Comment (kannon92) [reply go.sum:66]: As a OSS team I think the right approach would be to open this issue on their repo and maybe propose a fix to their code.  

I don't think forking and including your fork is a pattern we want to encourage from the open source team.

Comment (ClifHouck) [reply go.sum:66]: Generally I agree with you, but the context of this repo matters. 

I'm happy to move this fork into `armadaproject` or just move the two source files into `armadaproject/armada`. To reiterate: the root of my fork is not actively maintained, as in it hasn't had a commit for *years*. It's likely maintained by a single person who has effectively abandoned the project. The changes I've made here also totally change the semantics and functionality of the pool, so unlikely to be merged upstream.

File: internal/jobservice/events/client.go
Comment (ClifHouck) [reply internal/jobservice/events/client.go:1]: Addressed in 9d0633f

File: internal/jobservice/application.go (hunk 2 above)
Comment (ClifHouck) [reply internal/jobservice/application.go]: Addressed in [a465a04](https://github.com/armadaproject/armada/pull/2420/commits/a465a04f4cf23b8f8cebdaabc1c24c657d0a8afd)

File: internal/common/grpc/grpcpool/LICENSE [hunk 3]
```diff
@@ -0,0 +1,21 @@
+MIT License
```
Comment (kannon92) [inline internal/common/grpc/grpcpool/LICENSE:1]: I really don't know what to do about this.  

I believe Armada-Operator is Licensed as Apache.  How does mixing different Software licenses work?

File: internal/jobservice/application.go [hunk 4]
```diff
@@ -30,12 +34,47 @@ func New() *App {
...
+var DefaultConfiguration = &configuration.JobServiceConfiguration{
+	GrpcPool: grpcconfig.GrpcPoolConfig{
+		InitialConnections: 5,
+		Capacity:           5,
+	},
+}
+
+// Mutates config where possible to correct mis-configurations.
+// Returns a non-nil error if mis-configuration is unrecoverable.
+func RectifyConfig(config *configuration.JobServiceConfiguration) error {
```
Comment (kannon92) [inline internal/jobservice/application.go:46]:
```suggestion
func RectifyConfig(config *configuration.JobServiceConfiguration) {
```

I don't think you ever return an error here.

File: internal/jobservice/application.go [hunk 5]
```diff
@@ -30,12 +34,47 @@ func New() *App {
...
+	return nil
+}
+
 func (a *App) StartUp(ctx context.Context, config *configuration.JobServiceConfiguration) error {
 	// Setup an errgroup that cancels on any job failing or there being no active jobs.
 	g, _ := errgroup.WithContext(ctx)
 
-	log := log.WithField("JobService", "Startup")
+	err := RectifyConfig(config)
+	if err != nil {
```
Comment (kannon92) [inline internal/jobservice/application.go:73]: I think this panic would never be hit.

**Review by kannon92 — COMMENTED** [review]
I think we should answer the mixing of licenses questions before we merge this.  

I'm not a lawyer but that screams an issue to me.  

The code LGTM though.  Just one nit about the fact that you don't return errors from the config function so we probably don't need the return type.

File: internal/common/grpc/grpcpool/LICENSE (hunk 3 above)
Comment (dave-gantenbein) [reply internal/common/grpc/grpcpool/LICENSE:1]: https://law.stackexchange.com/questions/6081/can-i-bundle-mit-licensed-components-in-a-apache-2-0-licensed-project#:~:text=Both%20the%20Apache%20License%20and,as%20required%20by%20the%20license.

Comment (kannon92) [reply internal/common/grpc/grpcpool/LICENSE:1]: well TIL.  

I guess its fine then?

Comment (dave-gantenbein) [reply internal/common/grpc/grpcpool/LICENSE:1]: Ship it.

File: internal/common/grpc/grpcpool/pool_test.go [hunk 6]
```diff
@@ -0,0 +1,282 @@
...
+			return grpc.Dial("example.com", grpc.WithInsecure())
+		}
+	}, 1, 1, 0)
+
+	if err != context.DeadlineExceeded {
+		t.Errorf("Returned error was not context.DeadlineExceeded, but the context was timed out before the initialization")
+	}
+}
+
+/* NOTE(Clif) No longer relevant
```
Comment (dave-gantenbein) [inline internal/common/grpc/grpcpool/pool_test.go:228]: If its not relevant, lets remove it.

File: internal/jobservice/application.go (hunk 4 above)
Comment (ClifHouck) [reply internal/jobservice/application.go:46]: I don't, but in the future I imagine there'd be a need to. This is only checking config variables I added in this PR but eventually we should probably be doing some kind of validation of all configuration options.

File: internal/jobservice/application.go (hunk 5 above)
Comment (ClifHouck) [reply internal/jobservice/application.go:73]: It won't but returning an error from `RectifyConfig` in the future is intended to indicate an unrecoverable error with `config`. Just putting in the correct behavior now.

File: internal/common/grpc/grpcpool/pool_test.go (hunk 6 above)
Comment (ClifHouck) [reply internal/common/grpc/grpcpool/pool_test.go]: Addressed in 440ce3e

File: internal/jobservice/application.go (hunk 4 above)
Comment (kannon92) [reply internal/jobservice/application.go:46]: This seems to me to be YAGNI.  If we expand to do validation we could add returning errors but as of right now we don't need the return.  I'll resolve though if you think that best.

**Review by kannon92 — APPROVED** [review]

**Review by suprjinx — APPROVED** [review]

File: internal/common/grpc/grpcpool/pool.go [hunk 7]
```diff
@@ -0,0 +1,277 @@
...
+type Factory func() (*grpc.ClientConn, error)
+
+// FactoryWithContext is a function type creating a grpc client
+// that accepts the context parameter that could be passed from
+// Get or NewWithContext method.
+type FactoryWithContext func(context.Context) (*grpc.ClientConn, error)
+
+// Pool is the grpc client pool
+type Pool struct {
+	// clients         chan *ClientConn
```
Comment (richscott) [inline internal/common/grpc/grpcpool/pool.go:35]: Can this commented-out line be removed?

**Review by richscott — APPROVED** [review]
I noticed a line of old commented-out code, but otherwise don't want to hold this up - looks good.

File: internal/common/grpc/grpcpool/pool.go (hunk 7 above)
Comment (ClifHouck) [reply internal/common/grpc/grpcpool/pool.go]: Addressed in ad97bac

**Review by richscott — APPROVED** [review]

**Pull request merged**